aptos-move/move-examples/data_structures/sources/smart_tree.move

```diff
@@ -0,0 +1,948 @@
+/// Type of large-scale search trees.
+///
+/// It internally uses BTree to organize the search tree data structure for keys. Comparing with
+/// other common search trees like AVL or Red-black tree, a BTree node has more children, and packs
+/// more metadata into one node, which is more disk friendly (and gas friendly).
+
+module aptos_std::smart_tree {
+    use aptos_std::bucket_table::{Self, BucketTable};
+    use std::option::{Self, Option};
+    use std::vector;
+
+    const E_UNKNOWN: u64 = 0;
+    const E_TREE_NOT_EMPTY: u64 = 1;
+    const E_TREE_TOO_BIG: u64 = 2;
+    const E_END_ITERATOR: u64 = 3;
+
+    const NULL_INDEX: u64 = 0;
+
+    const DEFAULT_ORDER : u8 = 32;
+
+    struct Node has drop, store {
+        is_leaf: bool,
+        parent: u64,
+        children: vector<Child>,
+        prev: u64,
+        next: u64,
+    }
+
+    struct Child has copy, drop, store {
+        max_key: u64,
+        index: u64,
+    }
+
+    struct Iterator has copy, drop {
+        node_index: u64,
+        key_index: u64,
+        key: u64,
+    }
+
+    struct SmartTree<V> has store {
+        root: u64,
+        nodes: BucketTable<u64, Node>,
+        entries: BucketTable<u64, V>,
+        order: u8,
+        min_leaf_index: u64,
+        max_leaf_index: u64,
+    }
+
+    /////////////////////////////////
+    // Constructors && Destructors //
+    /////////////////////////////////
+
+    /// Returns a new SmartTree with the default configuration.
+    public fun new<V: store>(): SmartTree<V> {
+        new_with_order(DEFAULT_ORDER)
+    }
+
+    /// Returns a new SmartTree with the provided order (the maximum # of children a node can have).
+    public fun new_with_order<V: store>(order: u8): SmartTree<V> {
+        assert!(order >= 5, E_UNKNOWN);
+        let root_node = new_node(/*is_leaf=*/true, /*parent=*/NULL_INDEX);
+        let nodes = bucket_table::new(1);
+        let root_index = 1;
```

Comment:
0 is used by NULL_INDEX as special value.

I always create the root node even the tree is empty, this can reduce special handlings for empty tree in other function.